File: backend/routes/bugReports.js
const express = require('express');
const router = express.Router();
const nodemailer = require('nodemailer');
const db = require('../database');
const { requireRole } = require('../middleware/auth');

function getTransporter() {
  const cfg = db.prepare('SELECT * FROM azienda WHERE id=1').get();
  if (!cfg?.smtp_host || !cfg?.smtp_user) return null;
  return nodemailer.createTransport({
    host: cfg.smtp_host,
    port: cfg.smtp_port || 587,
    secure: cfg.smtp_secure === 1,
    auth: { user: cfg.smtp_user, pass: cfg.smtp_pass },
  });
}

function getAdminEmail() {
  const cfg = db.prepare('SELECT smtp_from, smtp_user FROM azienda WHERE id=1').get();
  return cfg?.smtp_from || cfg?.smtp_user || null;
}

const prioritaLabel = { BASSA: 'Bassa', MEDIA: 'Media', ALTA: 'Alta ⚠️' };

// POST / – crea nuova segnalazione
router.post('/', async (req, res) => {
  const { titolo, descrizione, pagina, priorita } = req.body;
  if (!titolo || !descrizione) return res.status(400).json({ error: 'titolo e descrizione obbligatori' });

  const prio = ['BASSA', 'MEDIA', 'ALTA'].includes(priorita) ? priorita : 'MEDIA';
  const result = db.prepare(
    'INSERT INTO bug_reports (titolo, descrizione, pagina, priorita) VALUES (?, ?, ?, ?)'
  ).run(titolo, descrizione, pagina || '', prio);

  // Notifica email (non bloccante se SMTP non configurato)
  try {
    const t = getTransporter();
    const dest = getAdminEmail();
    if (t && dest) {
      const html = `
        <div style="font-family:Arial,sans-serif;font-size:13px;color:#1e293b;max-width:600px">
          <h2 style="color:#dc2626">🐛 Nuova segnalazione bug</h2>
          <table style="border-collapse:collapse;width:100%">
            <tr><td style="padding:6px 12px;font-weight:600;width:120px">Titolo</td><td style="padding:6px 12px">${titolo}</td></tr>
            <tr style="background:#f8fafc"><td style="padding:6px 12px;font-weight:600">Pagina</td><td style="padding:6px 12px">${pagina || '—'}</td></tr>
            <tr><td style="padding:6px 12px;font-weight:600">Priorità</td><td style="padding:6px 12px">${prioritaLabel[prio]}</td></tr>
            <tr style="background:#f8fafc"><td style="padding:6px 12px;font-weight:600;vertical-align:top">Descrizione</td><td style="padding:6px 12px;white-space:pre-wrap">${descrizione}</td></tr>
          </table>
          <p style="margin-top:16px;font-size:12px;color:#64748b">Segnalazione #${result.lastInsertRowid} — ${new Date().toLocaleString('it-IT')}</p>
        </div>`;
      await t.sendMail({ from: dest, to: dest, subject: `[Bug] ${titolo}`, html });
    }
  } catch (_) {}

  res.json({ id: result.lastInsertRowid });
});

// GET / – lista tutte le segnalazioni
router.get('/', (req, res) => {
  const rows = db.prepare('SELECT * FROM bug_reports ORDER BY created_at DESC').all();
  res.json(rows);
});

// PATCH /:id/risolto – segna come risolto
router.patch('/:id/risolto', (req, res) => {
  const { id } = req.params;
  const row = db.prepare('SELECT id FROM bug_reports WHERE id=?').get(id);
  if (!row) return res.status(404).json({ error: 'Non trovato' });
  db.prepare("UPDATE bug_reports SET stato='RISOLTO', resolved_at=datetime('now') WHERE id=?").run(id);
  res.json({ ok: true });
});

// PATCH /:id/riapri – riapri
router.patch('/:id/riapri', (req, res) => {
  const { id } = req.params;
  db.prepare("UPDATE bug_reports SET stato='APERTO', resolved_at=NULL WHERE id=?").run(id);
  res.json({ ok: true });
});

// DELETE /:id  (solo ruoli amministrativi)
router.delete('/:id', requireRole('SUPERADMIN', 'OWNER', 'ADMIN'), (req, res) => {
  db.prepare('DELETE FROM bug_reports WHERE id=?').run(req.params.id);
  res.json({ ok: true });
});

module.exports = router;

File: src-tauri/src/routes/bug_reports.rs
//! /api/bug-reports — parità con routes/bugReports.js (segnalazioni).
//! Nota: la notifica email (best-effort in Node, no-op se SMTP non configurato)
//! è demandata al modulo email; qui resta non bloccante e non influisce sulla
//! risposta HTTP, quindi è omessa per la parità del payload.

use axum::{
    extract::{Path, State},
    routing::{get, patch},
    Json, Router,
};
use rusqlite::params;
use serde_json::{json, Value};

use crate::db::AppState;
use crate::error::{ApiError, ApiResult};
use crate::web::tenant_conn;

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/", get(list).post(create))
        .route("/:id/risolto", patch(risolto))
        .route("/:id/riapri", patch(riapri))
        .route("/:id", axum::routing::delete(remove))
}

async fn create(State(state): State<AppState>, Json(body): Json<Value>) -> ApiResult<Json<Value>> {
    let titolo = body.get("titolo").and_then(Value::as_str).unwrap_or("");
    let descrizione = body.get("descrizione").and_then(Value::as_str).unwrap_or("");
    if titolo.is_empty() || descrizione.is_empty() {
        return Err(ApiError::Status(
            axum::http::StatusCode::BAD_REQUEST,
            "titolo e descrizione obbligatori".into(),
        ));
    }
    let prio_in = body.get("priorita").and_then(Value::as_str).unwrap_or("");
    let prio = if ["BASSA", "MEDIA", "ALTA"].contains(&prio_in) { prio_in } else { "MEDIA" };
    let pagina = body.get("pagina").and_then(Value::as_str).unwrap_or("");

    let conn = tenant_conn(&state)?;
    let conn = conn.lock().unwrap();
    conn.execute(
        "INSERT INTO bug_reports (titolo, descrizione, pagina, priorita) VALUES (?, ?, ?, ?)",
        params![titolo, descrizione, pagina, prio],
    )?;
    Ok(Json(json!({ "id": conn.last_insert_rowid() })))
}

async fn list(State(state): State<AppState>) -> ApiResult<Json<Value>> {
    let conn = tenant_conn(&state)?;
    let conn = conn.lock().unwrap();
    let mut stmt = conn.prepare(
        "SELECT id, titolo, descrizione, pagina, priorita, stato, created_at, resolved_at
         FROM bug_reports ORDER BY created_at DESC",
    )?;
    let rows = stmt
        .query_map([], |r| {
            Ok(json!({
                "id": r.get::<_, i64>(0)?,
                "titolo": r.get::<_, Option<String>>(1)?,
                "descrizione": r.get::<_, Option<String>>(2)?,
                "pagina": r.get::<_, Option<String>>(3)?,
                "priorita": r.get::<_, Option<String>>(4)?,
                "stato": r.get::<_, Option<String>>(5)?,
                "created_at": r.get::<_, Option<String>>(6)?,
                "resolved_at": r.get::<_, Option<String>>(7)?,
            }))
        })?
        .collect::<Result<Vec<_>, _>>()?;
    Ok(Json(Value::Array(rows)))
}

async fn risolto(State(state): State<AppState>, Path(id): Path<i64>) -> ApiResult<Json<Value>> {
    let conn = tenant_conn(&state)?;
    let conn = conn.lock().unwrap();
    let exists: Option<i64> = conn
        .query_row("SELECT id FROM bug_reports WHERE id=?", params![id], |r| r.get(0))
        .ok();
    if exists.is_none() {
        return Err(ApiError::Status(axum::http::StatusCode::NOT_FOUND, "Non trovato".into()));
    }
    conn.execute(
        "UPDATE bug_reports SET stato='RISOLTO', resolved_at=datetime('now') WHERE id=?",
        params![id],
    )?;
    Ok(Json(json!({ "ok": true })))
}

async fn riapri(State(state): State<AppState>, Path(id): Path<i64>) -> ApiResult<Json<Value>> {
    let conn = tenant_conn(&state)?;
    let conn = conn.lock().unwrap();
    conn.execute(
        "UPDATE bug_reports SET stato='APERTO', resolved_at=NULL WHERE id=?",
        params![id],
    )?;
    Ok(Json(json!({ "ok": true })))
}

async fn remove(State(state): State<AppState>, Path(id): Path<i64>) -> ApiResult<Json<Value>> {
    let conn = tenant_conn(&state)?;
    let conn = conn.lock().unwrap();
    conn.execute("DELETE FROM bug_reports WHERE id=?", params![id])?;
    Ok(Json(json!({ "ok": true })))
}
